extensions/amp-video/1.0/video-wrapper.js

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {MIN_VISIBILITY_RATIO_FOR_AUTOPLAY} from '../../../src/video-interface';
+import {useEffect, useRef, useState} from '../../../src/preact';
+import {useResourcesNotify} from '../../../src/preact/utils';
+// Source for this constant is css/video-autoplay.css
+import {Deferred} from '../../../src/utils/promise';
+import {
+  EMPTY_METADATA,
+  parseFavicon,
+  parseOgImage,
+  parseSchemaImage,
+  setMediaSession,
+} from '../../../src/mediasession-helper';
+import {cssText as autoplayCss} from '../../../build/video-autoplay.css';
+
+/**
+ * @param {!VideoWrapperProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function VideoWrapper({
+  component: Component,
+  autoplay = false,
+  controls = false,
+  noaudio = false,
+  mediasession = true,
+  children,
+  // Prevent override in passthroughProps.
+  muted: unusedMuted,
+  ...passthroughProps
+}) {
+  useResourcesNotify();
+
+  const [muted, setMuted] = useState(autoplay);
+  const [metadata, setMetadata] = useState(null);
+  const [playing, setPlaying] = useState(false);
+  const [userInteracted, setUserInteracted] = useState(false);
+
+  const wrapperRef = useRef(null);
+  const playerRef = useRef(null);
+
+  const readyDeferredRef = useRef(new Deferred());
+
+  const play = () => {
+    readyDeferredRef.current.promise.then(() => {
+      playerRef.current.play();
+    });
+  };
+
+  const pause = () => {
+    readyDeferredRef.current.promise.then(() => {
+      playerRef.current.pause();
+    });
+  };
+
+  useEffect(() => {
+    if (!mediasession || !metadata || !playing) {
+      return;
+    }
+    setMediaSession(window, metadata, play, pause);
+  }, [mediasession, playing, metadata]);
+
+  return (
+    <div
+      ref={wrapperRef}
+      style={{position: 'relative', width: '100%', height: '100%'}}
+    >
+      <Component
+        ref={playerRef}
+        // Ensure that props controlled here are excluded from passthroughProps,
+        // see destructuring statement above.
+        muted={muted}
+        controls={controls && (!autoplay || userInteracted)}
+        onLoad={() => readyDeferredRef.current.resolve()}
```

Comment:
Two things here:
1. I mentioned above, but I think i'd actually prefer if this was `onCanPlay` - i.e. as close to the video tag as possible.
2. Do we need to restart the `readyDeffered` promise if the video's source is changed?

Reply:
1. See above, changed.
2. Probably, but I'm not sure what the most elegant way to do this would be. Components use different interfaces to set source: an attribute (`src`) or a combination of them (params added to query), or children (`<source>`). We might need a separate event handler like `onLoadStart`.

Reply:
Is `onLoadStart` consistently available? Because that'd seem like the best answer. Alternatively, we could use `onLayoutEffect` and try to catch all nuances that could change source and thus loading.